runtest.py

```diff
@@ -258,31 +258,42 @@ def flush(self, data):
     sys.stderr = Tee(logfile, sys.stderr)
 
 # --- define helpers ----
-if sys.platform in ('win32', 'cygwin'):
-
-    def whereis(file):
-        pathext = [''] + os.environ['PATHEXT'].split(os.pathsep)
-        for d in os.environ['PATH'].split(os.pathsep):
-            f = os.path.join(d, file)
-            for ext in pathext:
-                fext = f + ext
-                if os.path.isfile(fext):
-                    return fext
-        return None
-
-else:
+if sys.platform == 'win32':
+    # thanks to Eryk Sun for this recipe
+    import ctypes
+
+    shlwapi = ctypes.OleDLL('shlwapi')
+    shlwapi.AssocQueryStringW.argtypes = (
+        ctypes.c_ulong,  # flags
+        ctypes.c_ulong,  # str
+        ctypes.c_wchar_p,  # pszAssoc
+        ctypes.c_wchar_p,  # pszExtra
+        ctypes.c_wchar_p,  # pszOut
+        ctypes.POINTER(ctypes.c_ulong),  # pcchOut
+    )
 
-    def whereis(file):
-        for d in os.environ['PATH'].split(os.pathsep):
-            f = os.path.join(d, file)
-            if os.path.isfile(f):
-                try:
-                    st = os.stat(f)
-                except OSError:
-                    continue
-                if stat.S_IMODE(st[stat.ST_MODE]) & 0o111:
-                    return f
-        return None
+    ASSOCF_NOTRUNCATE = 0x00000020
+    ASSOCF_INIT_IGNOREUNKNOWN = 0x00000400
+    ASSOCSTR_COMMAND = 1
+    ASSOCSTR_EXECUTABLE = 2
+    E_POINTER = ctypes.c_long(0x80004003).value
+
+    def get_template_command(filetype, verb=None):
+        flags = ASSOCF_INIT_IGNOREUNKNOWN | ASSOCF_NOTRUNCATE
+        assoc_str = ASSOCSTR_COMMAND
+        cch = ctypes.c_ulong(260)
+        while True:
```

Comment:
Why `while True` ?

Reply:
Beats me. I got an updated snip just now I'll check if it still does that.